rust/s3heap-service/src/lib.rs

```diff
@@ -0,0 +1,535 @@
+    /// # Returns
+    /// The config object.
+    /// # Panics
+    /// - If the config file cannot be read.
+    /// - If the config file is not valid YAML.
+    /// - If the config file does not contain the required fields.
+    /// - If the config file contains invalid values.
+    /// - If the environment variables contain invalid values.
+    /// # Notes
+    /// The environment variables are prefixed with CHROMA_ and are uppercase.
+    /// Values in the envionment variables take precedence over values in the YAML file.
+    // NOTE:  Copied to ../load/src/config.rs.
+    pub fn load_from_path(path: &str) -> Self {
+        println!("loading config from {path}");
+        println!(
+            r#"Full config is:
+================================================================================
+{}
+================================================================================
+"#,
+            std::fs::read_to_string(path)
+                .expect("should be able to open and read config to string")
+        );
+        // Unfortunately, figment doesn't support environment variables with underscores. So we have to map and replace them.
+        // Excluding our own environment variables, which are prefixed with CHROMA_.
+        let mut f = figment::Figment::from(Env::prefixed("CHROMA_").map(|k| match k {
+            k if k == "my_member_id" => k.into(),
+            k => k.as_str().replace("__", ".").into(),
+        }));
+        if std::path::Path::new(path).exists() {
+            f = figment::Figment::from(Yaml::file(path)).merge(f);
+        }
+        // Apply defaults - this seems to be the best way to do it.
+        // https://github.com/SergioBenitez/Figment/issues/77#issuecomment-1642490298
+        // f = f.join(Serialized::default(
+        //     "worker.num_indexing_threads",
+        //     num_cpus::get(),
+        // ));
+        let res = f.extract();
+        match res {
+            Ok(config) => config,
+            Err(e) => panic!("Error loading config: {}", e),
+        }
+    }
+}
+
+/// Configuration for OpenTelemetry tracing and observability.
+///
+/// Controls how the service exports traces and metrics to an OpenTelemetry collector.
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+pub struct OpenTelemetryConfig {
+    /// The OpenTelemetry collector endpoint URL.
+    #[serde(default = "OpenTelemetryConfig::default_endpoint")]
+    pub endpoint: String,
+    /// The service name reported in traces.
+    #[serde(default = "OpenTelemetryConfig::default_otel_service_name")]
+    pub service_name: String,
+    /// Trace level filters for different crates.
+    #[serde(default = "OpenTelemetryConfig::default_otel_filters")]
+    pub filters: Vec<OtelFilter>,
+}
+
+impl OpenTelemetryConfig {
+    fn default_endpoint() -> String {
+        "http://otel-collector:4317".to_string()
+    }
+
+    fn default_otel_service_name() -> String {
+        "rust-log-service".to_string()
```

Comment:
Assuming this is because we're running this on the RLS. Will traces from this module will still have some HeapTender specific prefixes?